pilot/pkg/model/push_context.go

```diff
@@ -1459,6 +1461,100 @@ func (ps *PushContext) initServiceRegistry(env *Environment) {
 	ps.initServiceAccounts(env, allServices)
 }
 
+// resolveServiceAliases sets the Aliases attributes on all services. The incoming Service's will just have AliasFor set,
+// but in our usage we often need the opposite: for a given service, what are all the aliases?
+// resolveServiceAliases walks this 'graph' of services and updates the Alias field in-place.
+func resolveServiceAliases(allServices []*Service, configsUpdated sets.Set[ConfigKey]) {
+	// rawAlias builds a map of Service -> AliasFor. So this will be ExternalName -> Service.
+	// In an edge case, we can have ExternalName -> ExternalName; we resolve that below
+	rawAlias := map[NamespacedHostname]host.Name{}
+	for _, s := range allServices {
+		if s.Resolution != Alias {
+			continue
+		}
+		nh := NamespacedHostname{
+			Hostname:  s.Hostname,
+			Namespace: s.Attributes.Namespace,
+		}
+		rawAlias[nh] = host.Name(s.Attributes.K8sAttributes.ExternalName)
+	}
+	// unnamespacedRawAlias is like rawAlias but without namespaces.
+	// This is because an `ExternalName` isn't namespaced. If there is a conflict, the behavior is undefined.
+	unnamespacedRawAlias := make(map[host.Name]host.Name, len(rawAlias))
+	for k, v := range rawAlias {
```

Comment:
nit: any reason this can't be collapsed into the previous loop?

Reply:
I think I did this to make it so the `make` can give the exact size